File: Task/TableViewCell/ActivityTableViewCell.m
//
//  ActivityTableViewCell.m
//  Task
//
//  Created by Sarankumar on 12/02/18.
//  Copyright © 2018 Saran. All rights reserved.
//

#import "ActivityTableViewCell.h"

@implementation ActivityTableViewCell

@synthesize titleLabel = _titleLabel;
@synthesize descriptionLabel = _descriptionLabel;
@synthesize displayImageView = _displayImageView;
@synthesize bottomLine = _bottomLine;

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupView];
    }
    return self;
}

- (void)setupView {
    // Title Label constraints added
    _titleLabel  = [[UILabel alloc] init];
    [self.contentView addSubview: _titleLabel];
    _titleLabel.numberOfLines = 0;
    _titleLabel.font = [UIFont systemFontOfSize: titleLabelFontSize weight: UIFontWeightBold];
    _titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    NSLayoutConstraint *tLleadingConstraint = [_titleLabel.leadingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.leadingAnchor constant: leadingConstant];
    NSLayoutConstraint *tLtrailingConstraint = [_titleLabel.trailingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.trailingAnchor constant: trailingConstant];
    NSLayoutConstraint *tLtopConstraint = [_titleLabel.topAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.topAnchor constant: topAndBottomConstant];
    // Display ImageView constraints added
    _displayImageView  = [[UIImageView alloc] init];
    [self.contentView addSubview: _displayImageView];
    _displayImageView.contentMode = UIViewContentModeScaleAspectFit;
    _displayImageView.translatesAutoresizingMaskIntoConstraints = NO;
    NSLayoutConstraint *iVleadingConstraint = [_displayImageView.leadingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.leadingAnchor constant: leadingConstant];
    NSLayoutConstraint *iVtrailingConstraint = [_displayImageView.trailingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.trailingAnchor constant: trailingConstant];
    NSLayoutConstraint *iVtopConstraint = [_displayImageView.topAnchor constraintEqualToAnchor:  _titleLabel.bottomAnchor constant: topAndBottomConstant];
    NSLayoutConstraint *iVHeightConstraint = [_displayImageView.heightAnchor constraintEqualToConstant: imageViewHeightConstant];
    // Description Label constraints added
    _descriptionLabel  = [[UILabel alloc] init];
    [self.contentView addSubview: _descriptionLabel];
    _descriptionLabel.numberOfLines = 0;
    _descriptionLabel.font = [UIFont systemFontOfSize: descriptionLabelFontSize weight: UIFontWeightRegular];
    _descriptionLabel.translatesAutoresizingMaskIntoConstraints = NO;
    NSLayoutConstraint *dLleadingConstraint = [_descriptionLabel.leadingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.leadingAnchor constant: leadingConstant];
    NSLayoutConstraint *dLtrailingConstraint = [_descriptionLabel.trailingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.trailingAnchor constant: trailingConstant];
    NSLayoutConstraint *dLtopConstraint = [_descriptionLabel.topAnchor constraintEqualToAnchor: _displayImageView.bottomAnchor constant: topAndBottomConstant];
    // Bottom View constraints added
    _bottomLine  = [[UIView alloc] init];
    [self.contentView addSubview: _bottomLine];
    _bottomLine.backgroundColor = [UIColor lightGrayColor];
    _bottomLine.translatesAutoresizingMaskIntoConstraints = NO;
    NSLayoutConstraint *bLleadingConstraint = [_bottomLine.leadingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.leadingAnchor constant: leadingConstant];
    NSLayoutConstraint *bLtrailingConstraint = [_bottomLine.trailingAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.trailingAnchor constant: trailingConstant];
    NSLayoutConstraint *bLtopConstraint = [_bottomLine.topAnchor constraintEqualToAnchor: _descriptionLabel.bottomAnchor constant: topAndBottomConstant];
    NSLayoutConstraint *bLbottomConstraint = [_bottomLine.bottomAnchor constraintEqualToAnchor: self.contentView.safeAreaLayoutGuide.bottomAnchor constant: bottomLineBottomConstant];
    NSLayoutConstraint *bLHeightConstraint = [_bottomLine.heightAnchor constraintEqualToConstant: bottomLineHeightConstant];
    
    [self.contentView addConstraints: @[tLleadingConstraint, tLtrailingConstraint, tLtopConstraint, iVleadingConstraint, iVtrailingConstraint, iVtopConstraint, iVHeightConstraint, dLleadingConstraint, dLtrailingConstraint, dLtopConstraint, bLleadingConstraint, bLtrailingConstraint, bLtopConstraint, bLbottomConstraint, bLHeightConstraint]];
}

//MARK:- Configure Cell
- (void)configureCell:(Activity *)activity {
    _titleLabel.text = activity.headLine;
    _descriptionLabel.text = activity.descriptionText;
    [_displayImageView sd_setImageWithURL:[NSURL URLWithString:activity.imageURL] placeholderImage:[UIImage imageNamed:@"Upload_Image"]];

}
@end

File: Task/TableViewCell/ActivityTableVIewCell.swift
//
//  ActivityTableVIewCell.swift
//  Task
//
//  Created by Sarankumar on 11/02/18.
//  Copyright © 2018 Saran. All rights reserved.
//

import Foundation
import UIKit
import SDWebImage

class ActivityTableVIewCell: UITableViewCell {
    private let leadingConstant: CGFloat = 20.0
    private let trailingConstant: CGFloat = -20.0
    private let topAndBottomConstant: CGFloat = 10.0
    private let imageViewHeightConstant: CGFloat = 100.0
    
    let profileImageView = UIImageView()
    let titleLabel = UILabel()
    let descriptionLabel = UILabel()
    let bottomLineView = UIView()

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.selectionStyle = .none
        self.setupView()
    }
    
    private func setupView() {
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        profileImageView.translatesAutoresizingMaskIntoConstraints = false
        descriptionLabel.translatesAutoresizingMaskIntoConstraints = false
        bottomLineView.translatesAutoresizingMaskIntoConstraints = false
        
        contentView.addSubview(titleLabel)
        contentView.addSubview(profileImageView)
        contentView.addSubview(descriptionLabel)
        contentView.addSubview(bottomLineView)
        
        titleLabel.font = UIFont.boldSystemFont(ofSize: 16.0)
        descriptionLabel.font = UIFont.systemFont(ofSize: 14.0)
        titleLabel.numberOfLines = 0
        descriptionLabel.numberOfLines = 0
        profileImageView.contentMode = .scaleAspectFit
        bottomLineView.backgroundColor = UIColor.bottomLine

        let safeView = contentView.safeAreaLayoutGuide
        
        let tLleadingConstraint    = titleLabel.leadingAnchor.constraint(equalTo: safeView.leadingAnchor, constant: leadingConstant)
        let tLtrailingConstraint   = titleLabel.trailingAnchor.constraint(equalTo: safeView.trailingAnchor, constant: trailingConstant)
        let tLtopConstraint        = titleLabel.topAnchor.constraint(equalTo: safeView.topAnchor, constant: topAndBottomConstant)
        
        let iVleadingConstraint    = profileImageView.leadingAnchor.constraint(equalTo: safeView.leadingAnchor, constant: leadingConstant)
        let iVtrailingConstraint   = profileImageView.trailingAnchor.constraint(equalTo: safeView.trailingAnchor, constant: trailingConstant)
        let iVtopConstraint        = profileImageView.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: topAndBottomConstant)
        let iVHeightConstraint     = profileImageView.heightAnchor.constraint(equalToConstant: imageViewHeightConstant)
        
        let dLleadingConstraint    = descriptionLabel.leadingAnchor.constraint(equalTo: safeView.leadingAnchor, constant: leadingConstant)
        let dLtrailingConstraint   = descriptionLabel.trailingAnchor.constraint(equalTo: safeView.trailingAnchor, constant: trailingConstant)
        let dLtopConstraint        = descriptionLabel.topAnchor.constraint(equalTo: profileImageView.bottomAnchor, constant: topAndBottomConstant)
        
        let bLleadingConstraint    = bottomLineView.leadingAnchor.constraint(equalTo: safeView.leadingAnchor, constant: leadingConstant)
        let bLtrailingConstraint   = bottomLineView.trailingAnchor.constraint(equalTo: safeView.trailingAnchor, constant: trailingConstant)
        let bLtopConstraint        = bottomLineView.topAnchor.constraint(equalTo: descriptionLabel.bottomAnchor, constant: topAndBottomConstant)
        let bLbottomConstraint     = bottomLineView.bottomAnchor.constraint(equalTo: safeView.bottomAnchor, constant: -1.0)
        let bLHeightConstraint     = bottomLineView.heightAnchor.constraint(equalToConstant: 1.0)
        
        contentView.addConstraints([tLleadingConstraint, tLtrailingConstraint, tLtopConstraint, iVleadingConstraint, iVtrailingConstraint, iVtopConstraint, iVHeightConstraint, dLleadingConstraint, dLtrailingConstraint, dLtopConstraint, bLleadingConstraint, bLtrailingConstraint, bLtopConstraint, bLbottomConstraint, bLHeightConstraint])
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func configureCell(object activity: Activity?) {
        self.titleLabel.text = activity?.title
        self.descriptionLabel.text = activity?.descriptionText
        if let url = activity?.icon {
            self.profileImageView.sd_setImage(with: URL(string: url), placeholderImage: UIImage(named: "Upload_Image"))
        } else {
             self.profileImageView.image = UIImage(named: "Upload_Image")
        }
    }
}
